buildSrc/subprojects/binary-compatibility/src/main/groovy/gradlebuild.binary-compatibility.gradle

```diff
@@ -84,55 +103,35 @@ dependencies {
-            if (subproject.plugins.hasPlugin("kotlin")) {
-                apiSourceFolders.from(subproject.sourceSets.main.kotlin.sourceDirectories)
-            }
-            newArchivesFiles.from(subproject.jar)
-            // TODO: Current classpath should be derived from the other "Gradle runtime" sources like api-metadata, runtime-api, distribution building, etc...
-            currentClasspath project(subproject.path)
-
-            // This transform uses the result of the exploded zip extraction
-            // and returns a single jar file based on the lookup project name
-            String projectName = subproject.moduleIdentity.baseName.get() - 'gradle-'
-            registerTransform(FindGradleJar) {
-                from.attribute(ARTIFACT_TYPE, 'gradle-libs-dir')
-                to.attribute(ARTIFACT_TYPE, projectName)
-                parameters {
-                    target = projectName
-                }
-            }
-            def appendixName = projectName.split('-')*.capitalize().join('')
-            def baselineConf = configurationsContainer.create("japicmp-baseline-${appendixName}") {
-                extendsFrom configurationsContainer.baseline
-                attributes.attribute(ARTIFACT_TYPE, projectName)
-            }
-            oldArchivesFiles.from(baselineConf)
-        }
-    }
 }
 
 def checkBinaryCompatibility = tasks.register("checkBinaryCompatibility", JapicmpTask) {
     def baseVersion = moduleIdentity.version.map { it.baseVersion.version }
     def isSnapshot = moduleIdentity.snapshot
-
-    inputs.files(configurations.currentClasspath)
     inputs.property('acceptedViolations', acceptedViolations.toAcceptedChangesMap())
     inputs.property("baseline.version", compatibilityBaselineVersion)
     inputs.property("currentVersion", baseVersion)
+    def apiSourceFolders = configurations.currentSources.incoming.artifactView { lenient(true) }.files
     inputs.files("apiSourceFolders", apiSourceFolders)
+    def currentClasspath = configurations.currentApiClasspath.incoming.artifactView { lenient(true) }.files
+    inputs.files(currentClasspath)
 
+    newClasspath = currentClasspath
     oldClasspath = configurations.baselineClasspath
-    newClasspath = configurations.currentClasspath
-    oldArchives = oldArchivesFiles
-    newArchives = newArchivesFiles
+
+    def currentDistributionJars = currentClasspath.filter { it.name.startsWith('gradle-') && it.name.endsWith('.jar') }
+    newArchives = currentDistributionJars
+    oldArchives = configurations.baselineJars
+
+    dependencies.registerTransform(FindGradleJars) {
+        from.attribute(ARTIFACT_TYPE, 'gradle-libs-dir')
+        to.attribute(ARTIFACT_TYPE, 'gradle-baseline-jars')
+        parameters {
+            currentJars.from(currentDistributionJars)
+            baselineVersion.set(compatibilityBaselineVersion)
+            currentVersion.set(baseVersion)
+        }
+    }
```

Comment:
@jjohannes I replaced the baseline-jar finding transforms that would be created for each file with this single one.
Does it still make `currentDistributionJars` resolve during configuration time? How can I debug to answer this question?

Reply:
I also considered an alternative here to write a task that would produce the required jars by extracting them from the archive and cross-checking with the jars from the `currentDistributionJars` configuration. Would that work better?

Reply:
I think the solution you have now is great. Transforms are for this kind of use cases.

We could maybe use a task. But I think it is more complicated to set up. Transforms and tasks are not so different. Both perform some action. It's more the way they are wired and scheduled that is different.

Transforms are for processing files retrieved through dependency resolution. So they fit well here.

This code now does not resolve at configuration time. You don't use any eager methods on Configuration that triggers resolving anymore.

The transform itself is execution code that is called "during resolution". So inside the transform you can resolve/access other files.

There should be a warning here in the build scan if you resolve at resolution time:
https://ge.gradle.org/s/gcvp6og52szyi/performance/suggestions?open=WyJub25GaW5hbEdyYWRsZVZlcnNpb25DaGVjayIsIm5vbkNhY2hlYWJsZVRhc2tPdXRwdXQiXQ